src/dbnode/storage/mem_tracker.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package storage
+
+import "sync"
+
+type memoryTracker struct {
+	sync.Mutex
+
+	opts MemoryTrackerOptions
+
+	numLoadedBytes        int
+	numPendingLoadedBytes int
+
+	waitForDecWg *sync.WaitGroup
+}
+
+func (m *memoryTracker) IncNumLoadedBytes(x int) (okToLoad bool) {
+	m.Lock()
+	defer m.Unlock()
+	limit := m.opts.numLoadedBytesLimit
+	if limit == 0 {
+		// Limit of 0 means no limit.
+		return true
+	}
+	// This check is optimistic in the sense that as long as the number of loaded bytes
+	// is currently under the limit then x is accepted, regardless of how far over the
+	// limit it bring the value of numLoadedBytes.
+	//
+	// The reason for this is to avoid scenarios where some process gets permanently
+	// stuck because the amount of data it needs to load at once is larger than the limit
+	// and as a result its never able to make any progress.
+	//
+	// In practice this implementation should be fine for the vast majority of configurations
+	// and workloads.
+	if m.numLoadedBytes < limit {
+		m.numLoadedBytes += x
+		return true
+	}
+
+	return false
+}
+
+func (m *memoryTracker) NumLoadedBytes() int {
+	m.Lock()
+	defer m.Unlock()
+	return m.numLoadedBytes
+}
+
+func (m *memoryTracker) MarkLoadedAsPending() {
+	m.Lock()
+	m.numPendingLoadedBytes = m.numLoadedBytes
+	m.Unlock()
+}
+
+func (m *memoryTracker) DecPendingLoadedBytes() {
+	m.Lock()
+	m.numLoadedBytes -= m.numPendingLoadedBytes
```

Comment:
it wasnt a bug the way I was using it, but in theory if someone misused it and called dec twice in a row it would do the wrong thing. Fixed it and added a unit test.